pkg/issuer/acme/setup_test.go

```diff
@@ -0,0 +1,520 @@
+					gen.SetIssuerConditionMessage(fmt.Sprintf(messageTemplateUpdateToV2, fmt.Sprintf("%s/", acmev1Staging), acmev2Staging))),
+			},
+		},
+		"ACME private key secret does not exist, account key generation not disabled, key secret creation fails": {
+			issuer: gen.IssuerFrom(baseIssuer,
+				gen.SetIssuerACMEPrivKeyRef(issuerSecretKeyName)),
+			kfsShouldBeCalled:          true,
+			kfsErr:                     notFoundErr,
+			acmePrivKeySecretCreateErr: someErr,
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorAccountRegistrationFailed),
+					gen.SetIssuerConditionMessage(messageAccountRegistrationFailed+someErr.Error())),
+			},
+			wantsErr: true,
+		},
+		"ACME private key secret does not exist, account key generation is disabled": {
+			issuer: gen.IssuerFrom(baseIssuer,
+				gen.SetIssuerACMEDisableAccountKeyGeneration(true),
+			),
+			kfsShouldBeCalled: true,
+			kfsErr:            notFoundErr,
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorAccountVerificationFailed),
+					gen.SetIssuerConditionMessage(messageAccountVerificationFailed+messageNoSecretKeyGenerationDisabled+notFoundErr.Error())),
+			},
+			wantsErr: true,
+		},
+		"ACME private key secret exists, but contains invalid private key": {
+			issuer:            gen.IssuerFrom(baseIssuer),
+			kfsShouldBeCalled: true,
+			kfsErr:            invalidDataErr,
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorAccountVerificationFailed),
+					gen.SetIssuerConditionMessage(fmt.Sprintf("%s%v", messageInvalidPrivateKey, invalidDataErr))),
+			},
+		},
+		"Checking ACME private key secret fails with an unknown error": {
+			issuer:            gen.IssuerFrom(baseIssuer),
+			kfsShouldBeCalled: true,
+			kfsErr:            someErr,
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorAccountVerificationFailed),
+					gen.SetIssuerConditionMessage(messageAccountVerificationFailed+someErr.Error())),
+			},
+			wantsErr: true,
+		},
+		"ACME account's key is not an RSA key": {
+			issuer: gen.IssuerFrom(baseIssuer,
+				gen.SetIssuerACMEPrivKeyRef(issuerSecretKeyName)),
+			kfsShouldBeCalled: true,
+			kfsKey:            ecdsaPrivKey,
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorAccountVerificationFailed),
+					gen.SetIssuerConditionMessage(fmt.Sprintf(messageTemplateNotRSA, issuerSecretKeyName))),
+			},
+		},
+		"ACME server URL is an invalid URL": {
+			issuer: gen.IssuerFrom(baseIssuer,
+				gen.SetIssuerACMEURL(invalidURL)),
+			kfsShouldBeCalled:          true,
+			kfsKey:                     rsaPrivKey,
+			removeClientShouldBeCalled: true,
```

Comment:
Why should remove client be called in this particular case?

Reply:
It's called once every time `Setup` function gets past URL check and ensuring private key is up to date.
New client is added to the cache only before exiting this func with 'success' (so it can be used by challenge and orders controllers as I understand) [here](https://github.com/jetstack/cert-manager/blob/master/pkg/issuer/acme/setup.go#L173) or [here](https://github.com/jetstack/cert-manager/blob/master/pkg/issuer/acme/setup.go#L274).

Do you mean that we should remove the client earlier?

Reply:
> Do you mean that we should remove the client earlier?

Yes. I'd have expected either that the ACME client would be removed right at the start of the Setup function and then re-added when Setup exits successfully OR I'd expect that the existing ACME client be left in place until the Setup function exits and at that point removed / replaced depending on whether Setup exits with an error.

We should start by documenting the desired behaviour in the Setup doc comments, but I keep forgetting that this PR is a first step and is about testing existing behavour. The documentation and refactoring can be added in later PRs.

Reply:
I noticed a comment on registry's `ListClients` [here](https://github.com/jetstack/cert-manager/blob/master/pkg/acme/accounts/registry.go#L58) suggesting that we should clean up the account registry when issuers are deleted (we don't currently do that), also see @munnerz 's comment on #2907 PR which implements the accounts registry:

> This PR does not change how we insert and remove clients from the cache, and this is currently still done in pkg/issuer/acme/setup.go. I am going to follow up with some work done to the 'issuer' controller(s) to have them properly Remove/Add clients in response to issuers being updated/deleted.

Perhaps we should do the mentioned follow-up work and as part of that we could also move up where the client is removed in `Setup` . I have created #4004 for that.